### ISSUE TITLE
DSP inside scenes behaves weirdly

### ISSUE THREAD

**Issue by salkin-mada** [eirikblekesaune/VTMax#33]
The DSP symbol on the Scene view / Scene Patch inside the Scene Core is still ON, but dsp is not working inside that core until the dsp is switched off on..
this happens over time, is the experience.
open a scene that uses dsp in the core, keep the scene activated. go to lunch (what not) then suddenly dsp stops working in the core.

**Comment by salkin-mada** [comment]
here is a reproducer:

{
	"patcher" : 	{
		"fileversion" : 1,
		"appversion" : 		{
			"major" : 8,
			"minor" : 0,
			"revision" : 6,
			"architecture" : "x64",
			"modernui" : 1
		}
,
		"classnamespace" : "box",
		"rect" : [ 59.0, 104.0, 644.0, 767.0 ],
		"bglocked" : 0,
		"openinpresentation" : 0,
		"default_fontsize" : 12.0,
		"default_fontface" : 0,
		"default_fontname" : "Arial",
		"gridonopen" : 1,
		"gridsize" : [ 15.0, 15.0 ],
		"gridsnaponopen" : 1,
		"objectsnaponopen" : 1,
		"statusbarvisible" : 2,
		"toolbarvisible" : 1,
		"lefttoolbarpinned" : 0,
		"toptoolbarpinned" : 0,
		"righttoolbarpinned" : 0,
		"bottomtoolbarpinned" : 0,
		"toolbars_unpinned_last_save" : 0,
		"tallnewobj" : 0,
		"boxanimatetime" : 200,
		"enablehscroll" : 1,
		"enablevscroll" : 1,
		"devicewidth" : 0.0,
		"description" : "",
		"digest" : "",
		"tags" : "",
		"style" : "",
		"subpatcher_template" : "",
		"boxes" : [ 			{
				"box" : 				{
					"fontface" : 0,
					"fontname" : "Arial",
					"fontsize" : 12.0,
					"id" : "obj-16",
					"maxclass" : "number~",
					"mode" : 2,
					"numinlets" : 2,
					"numoutlets" : 2,
					"outlettype" : [ "signal", "float" ],
					"patching_rect" : [ 442.75, 618.0, 56.0, 22.0 ],
					"sig" : 0.0
				}

			}
, 			{
				"box" : 				{
					"fontface" : 0,
					"fontname" : "Arial",
					"fontsize" : 12.0,
					"id" : "obj-15",
					"maxclass" : "number~",
					"mode" : 2,
					"numinlets" : 2,
					"numoutlets" : 2,
					"outlettype" : [ "signal", "float" ],
					"patching_rect" : [ 331.75, 618.0, 56.0, 22.0 ],
					"sig" : 0.0
				}

			}
, 			{
				"box" : 				{
					"fontface" : 0,
					"fontname" : "Arial",
					"fontsize" : 12.0,
					"id" : "obj-6",
					"maxclass" : "number~",
					"mode" : 2,
					"numinlets" : 2,
					"numoutlets" : 2,
					"outlettype" : [ "signal", "float" ],
					"patching_rect" : [ 242.0, 618.0, 56.0, 22.0 ],
					"sig" : 0.0
				}

			}
, 			{
				"box" : 				{
					"fontface" : 0,
					"fontname" : "Arial",
					"fontsize" : 12.0,
					"id" : "obj-5",
					"maxclass" : "number~",
					"mode" : 2,
					"numinlets" : 2,
					"numoutlets" : 2,
					"outlettype" : [ "signal", "float" ],
					"patching_rect" : [ 144.5, 618.0, 56.0, 22.0 ],
					"sig" : 0.0
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-109",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 170.0, 565.0, 34.0, 22.0 ],
					"text" : "*~ 2."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-107",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 253.0, 565.0, 34.0, 22.0 ],
					"text" : "*~ 2."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-106",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 345.0, 565.0, 34.0, 22.0 ],
					"text" : "*~ 2."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-105",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 428.0, 565.0, 34.0, 22.0 ],
					"text" : "*~ 2."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-104",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 253.0, 499.0, 29.5, 22.0 ],
					"text" : "*~"
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-103",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 345.0, 504.0, 29.5, 22.0 ],
					"text" : "*~"
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-102",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 427.0, 504.0, 29.5, 22.0 ],
					"text" : "*~"
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-100",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 172.25, 504.0, 29.5, 22.0 ],
					"text" : "*~"
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-18",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 349.0, 337.0, 36.0, 22.0 ],
					"text" : "!-~ 1."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-19",
					"maxclass" : "newobj",
					"numinlets" : 2,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 417.0, 337.0, 36.0, 22.0 ],
					"text" : "!-~ 1."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-53",
					"maxclass" : "newobj",
					"numinlets" : 1,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 248.5, 206.0, 44.0, 22.0 ],
					"text" : "sig~ 0."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-52",
					"maxclass" : "newobj",
					"numinlets" : 1,
					"numoutlets" : 1,
					"outlettype" : [ "signal" ],
					"patching_rect" : [ 354.0, 206.0, 44.0, 22.0 ],
					"text" : "sig~ 0."
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-26",
					"maxclass" : "live.dial",
					"numinlets" : 1,
					"numoutlets" : 2,
					"outlettype" : [ "", "float" ],
					"parameter_enable" : 1,
					"patching_rect" : [ 248.5, 140.0, 41.0, 48.0 ],
					"saved_attribute_attributes" : 					{
						"valueof" : 						{
							"parameter_shortname" : "live.dial",
							"parameter_type" : 0,
							"parameter_unitstyle" : 1,
							"parameter_longname" : "live.dial[1]",
							"parameter_mmax" : 1.0
						}

					}
,
					"varname" : "live.dial[1]"
				}

			}
, 			{
				"box" : 				{
					"id" : "obj-25",
					"maxclass" : "live.dial",
					"numinlets" : 1,
					"numoutlets" : 2,
					"outlettype" : [ "", "float" ],
					"parameter_enable" : 1,
					"patching_rect" : [ 334.0, 139.0, 41.0, 48.0 ],
					"saved_attribute_attributes" : 					{
						"valueof" : 						{
							"parameter_shortname" : "live.dial",
							"parameter_type" : 0,
							"parameter_unitstyle" : 1,
							"parameter_longname" : "live.dial",
							"parameter_mmax" : 1.0
						}

					}
,
					"varname" : "live.dial"
				}

			}
 ],
		"lines" : [ 			{
				"patchline" : 				{
					"destination" : [ "obj-109", 0 ],
					"source" : [ "obj-100", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-105", 0 ],
					"source" : [ "obj-102", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-106", 0 ],
					"source" : [ "obj-103", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-107", 0 ],
					"source" : [ "obj-104", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-16", 0 ],
					"source" : [ "obj-105", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-15", 0 ],
					"source" : [ "obj-106", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-6", 0 ],
					"source" : [ "obj-107", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-5", 0 ],
					"source" : [ "obj-109", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-102", 1 ],
					"order" : 0,
					"source" : [ "obj-18", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-103", 1 ],
					"order" : 1,
					"source" : [ "obj-18", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-102", 0 ],
					"order" : 0,
					"source" : [ "obj-19", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-104", 0 ],
					"order" : 1,
					"source" : [ "obj-19", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-52", 0 ],
					"source" : [ "obj-25", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-53", 0 ],
					"source" : [ "obj-26", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-100", 0 ],
					"order" : 2,
					"source" : [ "obj-52", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-103", 0 ],
					"order" : 1,
					"source" : [ "obj-52", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-19", 0 ],
					"order" : 0,
					"source" : [ "obj-52", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-100", 1 ],
					"order" : 2,
					"source" : [ "obj-53", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-104", 1 ],
					"order" : 1,
					"source" : [ "obj-53", 0 ]
				}

			}
, 			{
				"patchline" : 				{
					"destination" : [ "obj-18", 0 ],
					"order" : 0,
					"source" : [ "obj-53", 0 ]
				}

			}
 ],
		"parameters" : 		{
			"obj-25" : [ "live.dial", "live.dial", 0 ],
			"obj-26" : [ "live.dial[1]", "live.dial", 0 ],
			"parameterbanks" : 			{

			}

		}
,
		"dependency_cache" : [  ],
		"autosave" : 0
	}

}

**Comment by eirikblekesaune** [comment]
So the audio inside a scene stops working. One theory is that it can be something that happens with the poly~ object that encapsulates the audio graph of the scene. If everything else works, apart from the audio, this could be the culprit.

When a scene audio has stopped, do you have to toggle the global DSP state, or is it enough to toggle the scene state?

Could you try to make a reproducer patch for this? I.e. something that is just a poly~ with an audio process running inside it, and then leave for lunch? Also, look if there is an AppNap  issue with Max, or screen saver, computer sleep etc.

Again here you could test if the global DSP needs to be toggled, or if it is enough to toggle the mue state for the poly~.

**Comment by blacksound** [comment]
@salkin-mada 
The reproducer, could you instead post it on a Gist on Pastebin? Or, if you want to use Github comment for this, please wrap it into code (triple backticks) formatting so it is easier to copy, and that something is not lost?
for example:

```
if (isAwesome){
  return true
}
```

**Comment by blacksound** [comment]
The reprodcer that you posted does't give much info about what causes the bug. As far as I can see it is only an audio patch, no? Could you make a test patch that wraps it into a poly~ for instance?

**Comment by salkin-mada** [comment]
yes yes

**Comment by salkin-mada** [comment]
Thanks for the quick response. I will elaborate as soon as there is time.

**Comment by salkin-mada** [comment]
@blacksound I have to toggle the global DSP state to make it work again.
I will come back with a proper example so that it is possible to test collectively.

**Comment by salkin-mada** [comment]
Dear all this is not exactly a test patch or reproducer as requested by @blacksound 
But for now I have committed the patch I reproduce the bug with here at Alna.
https://github.com/blacksound/VTMax/tree/master/misc/Scener/DSPFailReproducer

But good idea to start with a ~poly and go from there. Thanks Eirik. 
Will also check the AppNap. Screensaver and sleep is disabled here.

**Comment by balintlaczko** [comment]
Hey Guys! Ouch... This will be a tough one. I can only help you substantially when I get back to Oslo - which is 20th of August -, as I left the project macbook there, so i am only on windows at the moment. I am just looking through the patch without having the VT package (God, pls compile jamoma well for windows...) and I see there are some number~ objects in the core. Previously I have had freezing dsp issues with number~-s just like you described, over long time spans, so I would advise to strictly avoid using them (for patches which need to run for more than 30 minutes, LOL). There was a similar issue with snapshot~ for which I found a hack to surely avoid it: don't specify rate in the object, instead just bang it with another metro, and it won't freeze. I also found that average~ can produce the same freezing problem, though I've heard rumors that they fixed that.
**Bottom line: avoid using number~, snapshot~ and average~ for as long as you possibly can.**

One question though: when the freezing happens, is the UI also frozen/acting weird?

(Miss you all, guys!)

**Comment by salkin-mada** [comment]
AHA ! I will try to work around that in the core of the Scene in question and see if it happens again.
@blacksound @balintlaczko @magnusbugge 
It may be that this issue have nothing to do with DSP or `~poly`, but a special case problem when using `number~` `snapshot~` `average~`, will test more and come back to you guys.
Thanks Eirik and Balint.

**Comment by eirikblekesaune** [comment]
Good insight @balintlaczko !
So there you have a set of test patches to make:
* One audio patch with `number~`
* One audio patch with `snapshot~` with specified rate
* One audio patch with `snapshot~` without specified rate
* One audio patch with `average~`
* One audio with none of the object above, but maybe a `cycle~` or similar.

Make them into:
* scenes
* patch wrapped in `poly~`
* simple patch not wrapped in `poly~`


That is 15 patches in total (3x5), it could lead to more insight into which conditions are causing the bug to happen.

**Comment by salkin-mada** [comment]
@balintlaczko the UI is working fine. but the slider objects showing data after DSP processing is for sure frozen due to the fact they don't get any data until the DSP is switched on again.

**Comment by balintlaczko** [comment]
@salkin-mada yeah, but that's okay I guess. I suspect that this will be the number~ thing but let's test and see. 
@blacksound  I will fire up a patch with number~-s now and wait. Until I get back I will do similar super-stripped-down tests with other dsp objects too. In theory they fixed a lot of these problems in Max 8, or at least that's what I believed.

BTW there is a reliable multichannel snapshot object in the spat5 library called spat5.snapshot~. You can specify the number of @channels and the /rate, and it spits out nice little arrays of the vals. I vaguely remember that Thibaut also made a stable spat5.average~ partly because of the freezing bug of the vanilla sibling. 
(NB he also made udp objects which don't drop packets like the vanilla ones, I mean Cycling should start paying him at some point...)

**Comment by balintlaczko** [comment]
BTW just to be clear, is this happening on Max 8 or 7?

I am running a number~ freeze test for about an hour now, it still works,but I just remembered that I already tested this months ago when Max 8 came out, and it wasn't freezing then. Anyway I let it run, and see what happens.

**Comment by salkin-mada** [comment]
@balintlaczko
Max 8, this repo.

**Comment by balintlaczko** [comment]
Okay the number~ object seems to be solid on windows 10 wasapi, it's been going on for like 4 hours without a problem. Will try it in poly~ too.

**Comment by salkin-mada** [comment]
There is the possibility that it is a user error, will try to reproduce in another Scene. :)

**Comment by salkin-mada** [comment]
Okay small update here. It seems to be enough to just open the core when the Scene DSP is "frozen". Then all `~msp` objects "wake up"

**Comment by balintlaczko** [comment]
Aha. That tells me that there is something with the poly~... Is it in @parallel mode? Also, forgot to ask before, but when the DSP freezes, does the audio also "get stuck" on the same sample or just goes silent?

**Comment by salkin-mada** [comment]
When this happened the last week during work, I did not use Max for sound just OSC and the silly max->msp->max processing of some data. So cant say if the sound goes silent or freezes at one sample. Will check **@parallel** on monday when i am back. Thanks Balint.
Also I haven't been able to reproduce this bug/error in any other Scenes yet. But then again it happens sometimes, sometimes not. I am open for suggestions like "It may be an user error".

**Comment by balintlaczko** [comment]
I'm thinking that it can be a logical bug too from our end. Can you monitor (print) what messages does the poly~ get until it gets frozen? Maybe it somehow gets muted or reinitialized behind the back of the jamoma switch at some point? (Is that possible somehow?) If it consistently never happens with other scenes yet often with this one, then I think it can't be a general problem like a buggy vanilla object or something, it has to do something with this scene's specific setup... And then I guess the @parallel has nothing to do with it.

**Comment by salkin-mada** [comment]
news news. 
This seems to be a problem with max 8. @balintlaczko just send a bug report to cycling74. 
msp -> max (dsp visualisation and all control rate conversion) only works as expected in 44.1 or harmonic ratios of that samplerate. this is somehow connected to any audio rate msp objects frequencies of 1000, or multiples of that.
and this seems to clutter control rate stuff in the long run. which results in something that looks like dsp is frozen in the scene core. but it is not! it is the control rate conversion objects that freezes.

**Comment by salkin-mada** [comment]
it seems that cycling stepped up and fixed the issue.
